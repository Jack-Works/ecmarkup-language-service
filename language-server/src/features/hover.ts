import {
    type CancellationToken,
    type Connection,
    type Hover,
    type HoverClientCapabilities,
    type HoverParams,
    MarkupKind,
    type ServerCapabilities,
    type WorkDoneProgressReporter,
} from 'vscode-languageserver'
import type { TextDocument } from 'vscode-languageserver-textdocument'
import { word_at_cursor } from '../parser/text.js'
import { documents } from '../server-shared.js'
import { getText } from '../utils/biblio.js'
import { formatDocument } from '../utils/format.js'
import type { Program } from '../workspace/program.js'

export class Hovers {
    constructor(public capabilities: HoverClientCapabilities | undefined) {}

    async hover(
        document: TextDocument,
        program: Program,
        params: HoverParams,
        _token?: CancellationToken,
        _workDoneProgress?: WorkDoneProgressReporter,
    ): Promise<Hover | undefined> {
        const source = document.getText()
        const offset = document.offsetAt(params.position)
        const sourceFile = program.getSourceFile(document)
        const biblio = await program.resolveBiblio(document.uri)

        const { word, isGrammar, isIntrinsic } = word_at_cursor(source, offset)

        const entry = biblio.find((entry) => {
            if (isGrammar) return entry.type === 'production' && word === entry.name
            else if (isIntrinsic) return entry.type === 'term' && word === entry.term.slice(1, -1)
            return word === getText(entry)
        })
        if (entry) {
            const contents = formatDocument(entry, this.capabilities?.contentFormat)
            if (!contents) return undefined
            return { contents }
        }

        {
            const local = sourceFile.localDefinedGrammars.filter((define) => define.name === word)
            if (local[0]) {
                return { contents: { kind: MarkupKind.PlainText, language: 'grammarkdown', value: local[0].summary } }
            }
        }
        {
            const local = sourceFile.localDefinedAbstractOperations.filter((define) => define.name === word)
            if (local[0]) {
                return { contents: { kind: MarkupKind.PlainText, language: 'ecmarkup', value: local[0].summary } }
            }
        }
        return undefined
    }

    static enable(
        features: ServerCapabilities<never>,
        connection: Connection,
        program: Program,
        capabilities: HoverClientCapabilities | undefined,
    ) {
        connection.onHover((params, token, workDoneProgress) => {
            const document = documents.get(params.textDocument.uri)
            if (!document) return null
            const hover = new Hovers(capabilities)
            return hover.hover(document, program, params, token, workDoneProgress)
        })
        features.hoverProvider = {}
    }
}
