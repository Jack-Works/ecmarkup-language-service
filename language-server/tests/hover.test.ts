import { expect, it } from 'vitest'
import { MarkupKind } from 'vscode-languageserver-types'
import { HoverProvider } from '../src/features/hover.js'
import { File } from './File.js'

const capabilities = { contentFormat: [MarkupKind.Markdown, MarkupKind.PlainText] }
it('hover on biblio entries', async () => {
    const hover = new HoverProvider(capabilities)
    const { document, textDocument, markers, program } = File.of`
        Expression${File.mark('expression')}
        Call${File.mark('call')}
        %AsyncFunction${File.mark('asyncFunction')}%
        Assert${File.mark('assert')}
    `
    for (const { desc, position, around } of markers) {
        const result = await hover.hover(document, program, { textDocument, position })
        expect([around, result]).toMatchSnapshot(desc)
    }
})

it('hover on local defined grammar', async () => {
    const hover = new HoverProvider(capabilities)
    const { document, textDocument, markers, program } = File.of`
        <emu-grammar type="definition">
            MoreOneNight :: "more" "one" "night"
        </emu-grammar>
        <emu-clause>
            |MoreOneNight${File.mark('ref')}|
        </emu-clause>
        <emu-grammar>
            MoreOneNight${File.mark('ref in grammar')}
        </emu-grammar>
        <emu-clause>
            MoreOneNight${File.mark('possible ref')}
        </emu-clause>
    `
    for (const { desc, position, around } of markers) {
        const result = await hover.hover(document, program, { textDocument, position })
        expect([around, result]).toMatchSnapshot(desc)
    }
})

it('hover on local defined AO', async () => {
    const hover = new HoverProvider(capabilities)
    const { document, textDocument, markers, program } = File.of`
        <emu-clause id="sec-SaveTheWorld" type="abstract operation">
            <h1>
                SaveTheWorld (
                    _input_: an ECMAScript language value,
                ): either a normal completion containing an Object or a throw completion
            </h1>
            <dl class="header">
            </dl>
            <emu-alg>
                1. TODO.
            </emu-alg>
        </emu-clause>
        <emu-alg>
            1. Perform ? SaveThe${File.mark()}World(  ).
        </emu-alg>
    `
    for (const { desc, position, around } of markers) {
        const result = await hover.hover(document, program, { textDocument, position })
        expect([around, result]).toMatchSnapshot(desc)
    }
})
