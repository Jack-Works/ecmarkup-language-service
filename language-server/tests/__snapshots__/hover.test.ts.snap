// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hover on biblio entries > assert 1`] = `
[
  "AssertðŸ”½",
  {
    "contents": {
      "kind": "markdown",
      "value": "[https://tc39.es/ecma262/#assert](https://tc39.es/ecma262/#assert)",
    },
  },
]
`;

exports[`hover on biblio entries > asyncFunction 1`] = `
[
  "%AsyncFunctionðŸ”½%",
  {
    "contents": {
      "kind": "markdown",
      "value": "[https://tc39.es/ecma262/#sec-async-function-constructor](https://tc39.es/ecma262/#sec-async-function-constructor)",
    },
  },
]
`;

exports[`hover on biblio entries > call 1`] = `
[
  "CallðŸ”½",
  {
    "contents": {
      "kind": "markdown",
      "value": "\`\`\`ts
Call(
  F: Value,
  V: Value,
  argumentsList?: List<Values>
): Completion<Value>
\`\`\`

[https://tc39.es/ecma262/#sec-call](https://tc39.es/ecma262/#sec-call)

This abstract operation may triggers user code.",
    },
  },
]
`;

exports[`hover on biblio entries > expression 1`] = `
[
  "ExpressionðŸ”½",
  {
    "contents": {
      "kind": "markdown",
      "value": "[https://tc39.es/ecma262/#prod-Expression](https://tc39.es/ecma262/#prod-Expression)",
    },
  },
]
`;

exports[`hover on local defined AO 1`] = `
[
  "1. Perform ? SaveTheðŸ”½World(  ).",
  {
    "contents": {
      "kind": "plaintext",
      "language": "ecmarkup",
      "value": "SaveTheWorld (
    _input_: an ECMAScript language value,
): either a normal completion containing an Object or a throw completion",
    },
  },
]
`;

exports[`hover on local defined grammar > possible ref 1`] = `
[
  "MoreOneNightðŸ”½",
  {
    "contents": {
      "kind": "plaintext",
      "language": "grammarkdown",
      "value": "MoreOneNight :: "more" "one" "night"",
    },
  },
]
`;

exports[`hover on local defined grammar > ref 1`] = `
[
  "|MoreOneNightðŸ”½|",
  {
    "contents": {
      "kind": "plaintext",
      "language": "grammarkdown",
      "value": "MoreOneNight :: "more" "one" "night"",
    },
  },
]
`;

exports[`hover on local defined grammar > ref in grammar 1`] = `
[
  "MoreOneNightðŸ”½",
  {
    "contents": {
      "kind": "plaintext",
      "language": "grammarkdown",
      "value": "MoreOneNight :: "more" "one" "night"",
    },
  },
]
`;
